Configuration/src/main/java/io/deephaven/configuration/Configuration.java

```diff
@@ -18,47 +21,142 @@
 
+    private static Configuration DEFAULT;
+    private static final Map<String, Configuration> NAMED_CONFIGURATIONS = new HashMap<>();
+
     // This should never be null to meet the contract for getContextKeyValues()
     private Collection<String> contextKeys = Collections.emptySet();
 
+    /**
+     * The default configuration implementation loading the property file from the default property file.
+     */
+    private static class DefaultConfiguration extends Configuration {
+        DefaultConfiguration() {
+            super();
+        }
+
+        @SuppressWarnings("unused")
+        public String getPropertyNullable(String propertyName) {
+            return properties.getProperty(propertyName);
+        }
+    }
+
+    /**
+     * A Named configuration that loads values from the file defined by the property `Configuration.name.rootFile`
+     */
+    private static class NamedConfiguration extends Configuration {
+        private final String name;
+
+        NamedConfiguration(final @NotNull String name) {
+            this.name = name;
+        }
+
+        @Override
+        String determinePropertyFile() {
+            final String propFile = System.getProperty("Configuration." + name + ".rootFile");
+            if (propFile == null) {
+                throw new ConfigurationException("Unable to load named configuration " + name);
+            }
+
+            return propFile;
+        }
+    }
+
     /**
      * Get the default Configuration instance.
      *
      * @return the single instance of Configuration allowed in an application
      */
     public static Configuration getInstance() {
-        if (INSTANCE == null) {
-            INSTANCE = new NullableConfiguration();
+        if (DEFAULT == null) {
+            DEFAULT = new DefaultConfiguration();
+            init(DEFAULT);
         }
-        return INSTANCE;
+        return DEFAULT;
     }
 
-    @SuppressWarnings("UnusedReturnValue")
-    public static Configuration newConfigurationForTesting() {
-        return new Configuration();
+    /**
+     * Get the {@link Configuration} for the specified name.  If a unique property file is not set using the
+     * `Configuration.name.rootFile` property then this will fall back to the Default configuration file.
+     *
+     * @param name the name of the configuration to load
+     * @return the named configuration, or the default if no named configuration was defined.
+     */
+    public static Configuration getNamedOrDefault(@NotNull final String name) {
+        if (DEFAULT_CONF_NAME.equals(name)) {
+            return getInstance();
+        }
+
+        Configuration instance = NAMED_CONFIGURATIONS.get(name);
+        if (instance != null) {
+            return instance;
+        }
+
+        // We could use a ConcurrentHashMap, but loading configuration can take significant time, and it
+        // seems wasteful to add contention to every Configuration.getInstance().
```

Comment:
So the actual "take time to load" should not happen that often, and you are still synced on the static class.  If you are worried you could load the configuration outside of the lock, and only insert it under the lock.  That has the disadvantage of maybe wanting to load two things at once if you have multiple threads hitting the same unloaded configuration w/o more complexity.

What you're doing is not actually safe, because the mutations of NAMED_CONFIGURATIONS in the lock does not mean that reading it outside the lock on line 95 is actually safe.